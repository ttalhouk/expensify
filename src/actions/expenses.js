import uuid from 'uuid';
import database from '../firebase/firebase';


export const addExpense = (expense) => {
  return {
    type: 'ADD_EXPENSE',
    expense
  };
}

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = {
      description,
      note,
      amount,
      createdAt
    }
    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    }).catch((e) => {
      console.log(e);
    });
  };
};

export const removeExpense = ({id}) => {
  return {
    type:'REMOVE_EXPENSE',
    id
  }
}

export const startRemoveExpense = ({id}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({id}));
    }).catch((e) => {
      console.log(e);
    });
  };
};

export const editExpense = (id, updates) => {
  return (
    {
      type: 'EDIT_EXPENSE',
      id,
      updates
    }
  );
};
export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    }).catch((e) => {
      console.log(e);
    });
  };
};

export const setExpenses = (expenses) => {
  return {
    type: 'SET_EXPENSES',
    expenses
  }
}

export const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value')
      .then((snapshot) => {
        const expenses = [];
        snapshot.forEach((expense) => {
            expenses.push({
              id: expense.key,
              ...expense.val()
            })
          });
        dispatch(setExpenses(expenses));
      }).catch((error) => {
        console.log(error);
      });
  }
}
