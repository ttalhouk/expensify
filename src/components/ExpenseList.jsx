import React from 'react';
import {connect} from 'react-redux';

import selectExpenses from '../selectors/expenses';

import ExpenseListItem from './ExpenseListItem.jsx';

export const ExpenseList = (props) => {
  return (
    <div>
      {
        props.expenses.length === 0 ? (<p>Add an expense to get started</p>): (
          props.expenses.map((expense) => {
            return <ExpenseListItem key={expense.id} expense={expense} />
          })
        )
      }
      {}
    </div>
  )
}

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  }
}

export default connect(mapStateToProps)(ExpenseList);
