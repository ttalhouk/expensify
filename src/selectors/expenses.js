import moment from 'moment';

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
  return expenses.filter((expense) => {
    // const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
    // const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
    const startDateMatch = startDate ? startDate.isSameOrBefore(moment(expense.createdAt), 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(moment(expense.createdAt), 'day') : true;
    const textMatch =  expense.description.toLowerCase().includes(text.toLowerCase());

    return textMatch && startDateMatch && endDateMatch;
  }).sort((exp1, exp2) => {
    if (sortBy === 'date') {
      return exp2.createdAt - exp1.createdAt;
    } else if (sortBy === 'amount'){
      return exp2.amount - exp1.amount;
    }
  })
}

export default getVisibleExpenses;
