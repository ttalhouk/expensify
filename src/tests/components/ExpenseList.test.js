import React from 'react';
import {shallow} from 'enzyme';

import {ExpenseList} from '../../components/ExpenseList.jsx';

import {expenses} from '../fixtures/expenses'

describe('<ExpenseList />',() => {
  it('should render the list of expenses correctly', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should render no items message if no expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
  })
})
