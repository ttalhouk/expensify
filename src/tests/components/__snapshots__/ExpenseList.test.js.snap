// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExpenseList /> should render no items message if no expenses 1`] = `
<div>
  <p>
    Add an expense to get started
  </p>
</div>
`;

exports[`<ExpenseList /> should render the list of expenses correctly 1`] = `
<div>
  <ExpenseListItem
    expense={
      Object {
        "amount": 50,
        "createdAt": 345600000,
        "description": "rent",
        "id": 1,
        "note": "",
      }
    }
    key="1"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 40,
        "createdAt": 0,
        "description": "water bill",
        "id": 2,
        "note": "",
      }
    }
    key="2"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 500,
        "createdAt": -345600000,
        "description": "gas bill",
        "id": 3,
        "note": "",
      }
    }
    key="3"
  />
</div>
`;
